mautic/mautic
# Various fixes with token in URL

**Please be sure you are submitting this against the _staging_ branch.**

[//]: # ( Please answer the following questions: )

| Q  | A
| --- | ---
| Bug fix? | y
| New feature? | 
| Automated tests included? |
| Related user documentation PR URL | 
| Related developer documentation PR URL | 
| Issues addressed (#s or URLs) | Fixes #8446
| BC breaks? | 
| Deprecations? | 

[//]: # ( Note that all new features should have a related user and/or developer documentation PR in their respective repositories. )

[//]: # ( Required: )
#### Description:
This PR resolved servera issues with tokens in url.

1, Noticed If we use in href token with modifier:

`<a href="{pagelink=1}?special={contactfield=email|true}&date={contactfield=email}">my link</a>`

froala change it to 

`<a href="{pagelink=1}?special={contactfield=email%7Ctrue}&date={contactfield=email}">my link</a>`

2, Pagelink token is note replaced with link to page

- `<a href="{pagelink=1}?website={contactfield=website}">website</a>` 

Now working versions are:

- `<a href="{pagelink=1}?website={contactfield=website}">go to page</a>`
- `<a href="{contactfield=website}?email={contactfield=email}">go to website</a>`

3. sanitizeUrlQuery doesn't work If query already have ? symbol


[//]: # ( As applicable: )
#### Steps to reproduce the bug:
1. Create email and add HTML 
`<a href="{pagelink=1}?website={contactfield=website}">got to link </a> <br /> 
<a href="{contactfield=website}">go to website</a>`
- add UTM tags to emal
- website in contact profile add as http://domain.tld/?some=example
2. Reopen HTML from wysiwyg  and you see encode `{contactfield=email%7Ctrue}` (first bug)
3. If you send email, these tokens are not like you expected
- `{pagelink=1}` is trackable, but redirect is 404 with `{pagelink=1}` in URL
- `{contactfield=email%7Ctrue}` is also not replaced with contact email address
- `{contactfield=website}` were created as redirect URL, but redirection is to wrong address with encoded symbols `http://domain.tld?some%3Dexample%3Futm_source%3Dtest&utm_medium=newsletter`. We expect `http://domain.tld?some%3DexampleF&utm_source=test&utm_medium=newsletter`

#### Steps to test this PR:
1. Load up [this PR](https://mautibox.com/7548)
2. Repeat all steps and see If token is replaced correctly and redirection works


<a href="https://gitpod.io/#https://github.com/mautic/mautic/pull/7548"><img src="https://gitpod.io/button/open-in-gitpod.svg"/></a>
